src/io.rs

```diff
@@ -1,51 +1,122 @@
 use pyo3::prelude::*;
 
 use image;
-
-#[cfg(feature = "libjpeg-turbo")]
-use turbojpeg::{Decompressor, Image, PixelFormat};
+use turbojpeg;
 
 // internal libs
 use crate::tensor::cv;
 
-// implementation function for libjpeg-turbo to load images
-#[cfg(feature = "libjpeg-turbo")]
-fn _read_image_jpeg_impl(
-    file_path: String,
-) -> Result<(Vec<u8>, Vec<i64>), Box<dyn std::error::Error>> {
-    // get the JPEG data
-    let jpeg_data = std::fs::read(file_path)?;
-
-    // initialize a Decompressor
-    let mut decompressor = Decompressor::new()?;
-
-    // read the JPEG header with image size
-    let header = decompressor.read_header(&jpeg_data)?;
-    let (width, height) = (header.width, header.height);
-
-    // prepare a storage for the raw pixel data
-    let mut pixels = vec![0; height * width * 3];
-    let image = Image {
-        pixels: pixels.as_mut_slice(),
-        width,
-        pitch: 3 * width, // we use no padding between rows
-        height,
-        format: PixelFormat::RGB,
-    };
-
-    // decompress the JPEG data
-    decompressor.decompress_to_slice(&jpeg_data, image)?;
-
-    // return the raw pixel data and shape
-    Ok((pixels, vec![height as i64, width as i64, 3]))
+#[pyclass]
+pub struct ImageSize {
+    #[pyo3(get)]
+    pub width: usize,
+    #[pyo3(get)]
+    pub height: usize,
+}
+
+#[pyclass]
+pub struct ImageDecoder {
+    pub decompressor: turbojpeg::Decompressor,
+}
+
+#[pyclass]
+pub struct ImageEncoder {
+    pub compressor: turbojpeg::Compressor,
+}
+
+#[pymethods]
+impl ImageEncoder {
+    #[new]
+    pub fn new() -> ImageEncoder {
+        let compressor = turbojpeg::Compressor::new().unwrap();
+        ImageEncoder { compressor }
+    }
+
+    pub fn encode(&mut self, data: &[u8], shape: [usize; 3]) -> Vec<u8> {
+        let image = turbojpeg::Image {
+            pixels: data,
+            width: shape[1],
+            pitch: 3 * shape[1],
+            height: shape[0],
+            format: turbojpeg::PixelFormat::RGB,
+        };
+
+        let jpeg_data = self.compressor.compress_to_vec(image);
+
+        jpeg_data.unwrap()
+    }
+}
+
+impl Default for ImageEncoder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[pymethods]
+impl ImageDecoder {
+    #[new]
+    pub fn new() -> ImageDecoder {
+        let decompressor = turbojpeg::Decompressor::new().unwrap();
+        ImageDecoder { decompressor }
+    }
+
+    pub fn read_header(&mut self, jpeg_data: &[u8]) -> ImageSize {
```

Comment:
This will be useful outside the class? (also, not sure if this suggestion works here)
```suggestion
    pub(self) fn read_header(&mut self, jpeg_data: &[u8]) -> ImageSize {
```

Reply:
i have some uses cases at work where i want to just check image sizes from the bytes stream and don't want to decode the whole stream